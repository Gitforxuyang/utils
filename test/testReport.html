
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">github.com/Gitforxuyang/utils/utils.go (83.8%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">package utils

import (
        "errors"
        "fmt"
        "go.mongodb.org/mongo-driver/bson/primitive"
        "reflect"
        "strconv"
        "strings"
        "time"
)

//示例
//        type Animal struct {
//                Age int `default:"1"`  简单的设置默认值
//                Time int `default:",NOW_SECOND()"`   设置函数默认值， NOW_SECOND()=秒 NOW_MILLI()=毫秒
//        }
//为实现了IEntity的struct设置默认值
func Default(entity interface{}) error <span class="cov8" title="1">{
        if reflect.ValueOf(entity).Kind() != reflect.Ptr </span><span class="cov8" title="1">{
                return errors.New("传入参数必须是指针类型")
        }</span>
        <span class="cov8" title="1">return setDefault(reflect.ValueOf(entity).Elem(), reflect.TypeOf(entity).Elem())</span>
}

//设置默认值函数
func setDefault(_value reflect.Value, _type reflect.Type) error <span class="cov8" title="1">{
        //不能设置基础类型的指针默认值
        if _type.Kind() != reflect.Struct </span><span class="cov0" title="0">{
                return errors.New("设置默认值的类型必须是struct")
        }</span>
        <span class="cov8" title="1">for i := 0; i &lt; _value.NumField(); i++ </span><span class="cov8" title="1">{
                if _value.Field(i).Kind() == reflect.Struct </span><span class="cov8" title="1">{
                        setDefault(_value.Field(i), _value.Field(i).Type())
                        continue</span>
                }
                <span class="cov8" title="1">if _value.Field(i).Kind() == reflect.Ptr </span><span class="cov8" title="1">{
                        if _value.Field(i).IsNil() </span><span class="cov8" title="1">{
                                return errors.New("指针类型的属性不能为nil")
                        }</span>
                        <span class="cov8" title="1">setDefault(_value.Field(i).Elem(), _value.Field(i).Type().Elem())
                        continue</span>
                }
                <span class="cov8" title="1">_default := _type.Field(i).Tag.Get("default")
                if _default == "" </span><span class="cov0" title="0">{
                        continue</span>
                }
                <span class="cov8" title="1">_defaultArr := strings.Split(_default, ",")
                if len(_defaultArr) == 1 </span><span class="cov8" title="1">{
                        _default = _defaultArr[0]
                }</span> else<span class="cov8" title="1"> {
                        //如果设置了预制函数则设置 NOW_SECOND()=当前时间秒数 NOW_MILLI()=当前时间毫秒数
                        switch _defaultArr[1] </span>{
                        case "NOW_SECOND()":<span class="cov8" title="1">
                                _value.Field(i).SetInt(time.Now().Unix())
                                break</span>
                        case "NOW_MILLI()":<span class="cov8" title="1">
                                _value.Field(i).SetInt(time.Now().UnixNano() / 1e6)
                                break</span>
                        default:<span class="cov0" title="0">
                                return errors.New("未知的内置默认函数")
                                break</span>
                        }
                        <span class="cov8" title="1">continue</span>
                }

                <span class="cov8" title="1">switch _value.Field(i).Kind() </span>{
                case reflect.Bool:<span class="cov8" title="1">
                        d, err := strconv.ParseBool(_default)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetBool(d)
                        break</span>
                case reflect.Int:<span class="cov8" title="1">
                        d, err := strconv.ParseInt(_default, 10, 32)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetInt(d)
                        break</span>
                case reflect.Int8:<span class="cov8" title="1">
                        d, err := strconv.ParseInt(_default, 10, 8)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetInt(d)
                        break</span>
                case reflect.Int16:<span class="cov8" title="1">
                        d, err := strconv.ParseInt(_default, 10, 16)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetInt(d)
                        break</span>
                case reflect.Int32:<span class="cov8" title="1">
                        d, err := strconv.ParseInt(_default, 10, 32)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetInt(d)
                        break</span>
                case reflect.Int64:<span class="cov8" title="1">
                        d, err := strconv.ParseInt(_default, 10, 64)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetInt(d)
                        break</span>
                case reflect.Uint:<span class="cov8" title="1">
                        d, err := strconv.ParseUint(_default, 10, 32)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetUint(d)
                        break</span>
                case reflect.Uint8:<span class="cov8" title="1">
                        d, err := strconv.ParseUint(_default, 10, 8)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetUint(d)
                        break</span>
                case reflect.Uint16:<span class="cov8" title="1">
                        d, err := strconv.ParseUint(_default, 10, 16)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetUint(d)
                        break</span>
                case reflect.Uint32:<span class="cov8" title="1">
                        d, err := strconv.ParseUint(_default, 10, 32)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetUint(d)
                        break</span>
                case reflect.Uint64:<span class="cov8" title="1">
                        d, err := strconv.ParseUint(_default, 10, 64)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetUint(d)
                        break</span>
                case reflect.Float32:<span class="cov8" title="1">
                        d, err := strconv.ParseFloat(_default, 10)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetFloat(d)
                        break</span>
                case reflect.Float64:<span class="cov8" title="1">
                        d, err := strconv.ParseFloat(_default, 10)
                        if err != nil </span><span class="cov0" title="0">{
                                return err
                        }</span>
                        <span class="cov8" title="1">_value.Field(i).SetFloat(d)
                        break</span>
                case reflect.String:<span class="cov8" title="1">
                        _value.Field(i).SetString(_default)
                        break</span>
                }
        }
        <span class="cov8" title="1">return nil</span>
}

//将长id转化为短id 类似于000000000000000062361472-&gt;62361472
func LongIdToSortId(longId string) (int64, error) <span class="cov8" title="1">{
        str := strings.TrimLeft(longId, "0")
        return strconv.ParseInt(str, 10, 64)
}</span>

func SortIdToLongId(sortId int64) string <span class="cov8" title="1">{
        return fmt.Sprintf("%024d", sortId)
}</span>

//字符串转化为objectId 如果字符串长度不够则自动补足24位
func StringToObjectId(strId string) (*primitive.ObjectID, error) <span class="cov8" title="1">{
        longId := fmt.Sprintf("%024s", strId)
        objId, err := primitive.ObjectIDFromHex(longId)
        if err != nil </span><span class="cov8" title="1">{
                return nil, err
        }</span>
        <span class="cov8" title="1">return &amp;objId, nil</span>
}
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
