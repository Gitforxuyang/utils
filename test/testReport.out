mode: set
github.com/Gitforxuyang/utils/utils.go:20.40,21.51 1 1
github.com/Gitforxuyang/utils/utils.go:24.2,24.82 1 1
github.com/Gitforxuyang/utils/utils.go:21.51,23.3 1 1
github.com/Gitforxuyang/utils/utils.go:28.65,30.36 1 1
github.com/Gitforxuyang/utils/utils.go:33.2,33.41 1 1
github.com/Gitforxuyang/utils/utils.go:165.2,165.12 1 1
github.com/Gitforxuyang/utils/utils.go:30.36,32.3 1 0
github.com/Gitforxuyang/utils/utils.go:33.41,34.47 1 1
github.com/Gitforxuyang/utils/utils.go:38.3,38.44 1 1
github.com/Gitforxuyang/utils/utils.go:45.3,46.21 2 1
github.com/Gitforxuyang/utils/utils.go:49.3,50.28 2 1
github.com/Gitforxuyang/utils/utils.go:68.3,68.33 1 1
github.com/Gitforxuyang/utils/utils.go:34.47,36.12 2 1
github.com/Gitforxuyang/utils/utils.go:38.44,39.31 1 1
github.com/Gitforxuyang/utils/utils.go:42.4,43.12 2 1
github.com/Gitforxuyang/utils/utils.go:39.31,41.5 1 1
github.com/Gitforxuyang/utils/utils.go:46.21,47.12 1 0
github.com/Gitforxuyang/utils/utils.go:50.28,52.4 1 1
github.com/Gitforxuyang/utils/utils.go:52.9,54.26 1 1
github.com/Gitforxuyang/utils/utils.go:65.4,65.12 1 1
github.com/Gitforxuyang/utils/utils.go:55.24,57.10 2 1
github.com/Gitforxuyang/utils/utils.go:58.23,60.10 2 1
github.com/Gitforxuyang/utils/utils.go:61.12,63.10 2 0
github.com/Gitforxuyang/utils/utils.go:69.21,71.18 2 1
github.com/Gitforxuyang/utils/utils.go:74.4,75.9 2 1
github.com/Gitforxuyang/utils/utils.go:76.20,78.18 2 1
github.com/Gitforxuyang/utils/utils.go:81.4,82.9 2 1
github.com/Gitforxuyang/utils/utils.go:83.21,85.18 2 1
github.com/Gitforxuyang/utils/utils.go:88.4,89.9 2 1
github.com/Gitforxuyang/utils/utils.go:90.22,92.18 2 1
github.com/Gitforxuyang/utils/utils.go:95.4,96.9 2 1
github.com/Gitforxuyang/utils/utils.go:97.22,99.18 2 1
github.com/Gitforxuyang/utils/utils.go:102.4,103.9 2 1
github.com/Gitforxuyang/utils/utils.go:104.22,106.18 2 1
github.com/Gitforxuyang/utils/utils.go:109.4,110.9 2 1
github.com/Gitforxuyang/utils/utils.go:111.21,113.18 2 1
github.com/Gitforxuyang/utils/utils.go:116.4,117.9 2 1
github.com/Gitforxuyang/utils/utils.go:118.22,120.18 2 1
github.com/Gitforxuyang/utils/utils.go:123.4,124.9 2 1
github.com/Gitforxuyang/utils/utils.go:125.23,127.18 2 1
github.com/Gitforxuyang/utils/utils.go:130.4,131.9 2 1
github.com/Gitforxuyang/utils/utils.go:132.23,134.18 2 1
github.com/Gitforxuyang/utils/utils.go:137.4,138.9 2 1
github.com/Gitforxuyang/utils/utils.go:139.23,141.18 2 1
github.com/Gitforxuyang/utils/utils.go:144.4,145.9 2 1
github.com/Gitforxuyang/utils/utils.go:146.24,148.18 2 1
github.com/Gitforxuyang/utils/utils.go:151.4,152.9 2 1
github.com/Gitforxuyang/utils/utils.go:153.24,155.18 2 1
github.com/Gitforxuyang/utils/utils.go:158.4,159.9 2 1
github.com/Gitforxuyang/utils/utils.go:160.23,162.9 2 1
github.com/Gitforxuyang/utils/utils.go:71.18,73.5 1 0
github.com/Gitforxuyang/utils/utils.go:78.18,80.5 1 0
github.com/Gitforxuyang/utils/utils.go:85.18,87.5 1 0
github.com/Gitforxuyang/utils/utils.go:92.18,94.5 1 0
github.com/Gitforxuyang/utils/utils.go:99.18,101.5 1 0
github.com/Gitforxuyang/utils/utils.go:106.18,108.5 1 0
github.com/Gitforxuyang/utils/utils.go:113.18,115.5 1 0
github.com/Gitforxuyang/utils/utils.go:120.18,122.5 1 0
github.com/Gitforxuyang/utils/utils.go:127.18,129.5 1 0
github.com/Gitforxuyang/utils/utils.go:134.18,136.5 1 0
github.com/Gitforxuyang/utils/utils.go:141.18,143.5 1 0
github.com/Gitforxuyang/utils/utils.go:148.18,150.5 1 0
github.com/Gitforxuyang/utils/utils.go:155.18,157.5 1 0
github.com/Gitforxuyang/utils/utils.go:169.51,172.2 2 1
github.com/Gitforxuyang/utils/utils.go:174.42,176.2 1 1
github.com/Gitforxuyang/utils/utils.go:179.66,182.16 3 1
github.com/Gitforxuyang/utils/utils.go:185.2,185.20 1 1
github.com/Gitforxuyang/utils/utils.go:182.16,184.3 1 1
github.com/Gitforxuyang/utils/utils.go:189.65,193.56 3 1
github.com/Gitforxuyang/utils/utils.go:197.2,198.17 2 1
github.com/Gitforxuyang/utils/utils.go:210.2,210.19 1 1
github.com/Gitforxuyang/utils/utils.go:193.56,195.3 1 1
github.com/Gitforxuyang/utils/utils.go:199.21,199.21 0 0
github.com/Gitforxuyang/utils/utils.go:200.21,201.35 1 1
github.com/Gitforxuyang/utils/utils.go:206.3,206.8 1 1
github.com/Gitforxuyang/utils/utils.go:207.10,208.20 1 1
github.com/Gitforxuyang/utils/utils.go:201.35,202.59 1 1
github.com/Gitforxuyang/utils/utils.go:202.59,204.5 1 1
github.com/Gitforxuyang/utils/utils.go:213.25,216.2 2 1
github.com/Gitforxuyang/utils/utils.go:218.31,222.2 3 1
